### PR TITLE
docs(fern): add EAGLE speculative decoding e2e tutorial

### PR DESCRIPTION
**Pull request by khazic**
## What
Add a comprehensive end-to-end tutorial for EAGLE speculative decoding, covering EAGLE-1, EAGLE-2, and EAGLE-3 drafter training, dataset preparation, and SGLang serving.

## Changelog
- New Fern doc page: `guides/speculative/eagle.mdx` in the `nightly` docs tree.
- Added the navigation entry in `nightly.yml`.
- Left released `v0.4` / `latest` docs unchanged.

## Tutorial Structure
- **Step 0**: Environment setup (Docker + SGLang)
- **Step 1**: EAGLE architecture explained (with diagram)
- **Step 2**: Dataset preparation (PerfectBlend / regeneration with target model)
- **Step 3**: EAGLE-3 training config + launch (complete YAML + field reference)
- **Step 4**: Training logs, checkpoint layout, resume
- **Step 5**: SGLang serving (curl examples with expected JSON output)
- **Step 6**: Bonus EAGLE-1 training + EAGLE-1 vs EAGLE-3 comparison
- Troubleshooting table

Style follows the existing [Gemma 4 VLM tutorial](https://github.com/NVIDIA-NeMo/Automodel/blob/main/fern/versions/nightly/pages/guides/vlm/gemma4.mdx) — concrete examples, sample inputs/outputs, beginner-friendly.

## Pre-checks
- [x] DCO sign-off
- [x] YAML parsed successfully
- [x] `git diff --check` passed
- [ ] `fern check` not run locally because the Fern CLI is not installed in this environment

### CODE REVIEW

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+  [system] You are a helpful assistant....
+  [user] What are the main differences between Python 2 and Python 3?...
+  [assistant] Here are the key differences between Python 2 and Python 3:
+
+1. **P...
+```
+
+<Warning>
+PerfectBlend uses a `conversations` column, but `ChatDataset` expects `messages`.
+Rename the column before training:
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/speculative/eagle.mdx:187]: Better way might be adding messages_column to ChatDataset / build_eagle3_dataloader. Non blocking for now.

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+  micro_batch_size: 1
+  grad_accumulation_steps: 4   # effective batch = 8 GPUs * 1 * 4 = 32
+  num_workers: 4
+  num_epochs: 1
+
+  # EAGLE-3 specific
+  ttt_steps: 4                 # TTT unroll depth (higher = better but slower)
+  draft_vocab_size: 32000      # smaller vocab = faster drafter
+
+  freeze_embeddings: true
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/speculative/eagle.mdx:293]: Does freeze emb do anything?

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+  lr: 2.0e-4
+  betas: [0.9, 0.95]
+  weight_decay: 0.0
+  warmup_ratio: 0.05           # 5% warmup
+  min_lr_ratio: 0.1
+
+checkpoint:
+  enabled: true
+  checkpoint_dir: ./outputs/eagle3_llama8b/checkpoints
+  model_save_format: safetensors
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/speculative/eagle.mdx:310]: iirc, do we support safetensors+consolidate format here?

Comment (HuiyingLi) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:310]: If not could we add a warning? Perhaps guard it in the code

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+
+optimizer:
+  lr: 1.0e-4
+  betas: [0.9, 0.95]
+  weight_decay: 0.0
+
+checkpoint:
+  enabled: true
+  checkpoint_dir: ./outputs/eagle1_llama8b/checkpoints
+  model_save_format: safetensors
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/speculative/eagle.mdx:571]: same here?

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx (hunk 1 above)
Comment (khazic) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:187]: Agreed — a `messages_column` param on `ChatDataset` / `build_eagle3_dataloader` would be cleaner. Will track separately.

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx (hunk 2 above)
Comment (khazic) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:293]: Yes — the drafter copies the target's embedding table at init (`copy_embeddings_from_target`), then `freeze_embeddings()` sets `requires_grad=False` on those weights so only the draft transformer layers and lm_head are trained. Updated the tutorial with an inline comment explaining this.

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx (hunk 3 above)
Comment (khazic) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:310]: Yes — both EAGLE-1 and EAGLE-3 recipes default to `model_save_format="safetensors"` + `save_consolidated=True` in `_build_checkpointer` (train_eagle3.py:269-272, train_eagle1.py:207-210). The save path goes through the standard `Checkpointer.save_model`, same as LLM recipes. Added a comment in the tutorial noting these are the defaults shown explicitly for clarity.

Comment (khazic) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:310]: Confirmed it works — safetensors + consolidated is the hardcoded default in both recipes. No guard needed.

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx (hunk 4 above)
Comment (khazic) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:571]: Same answer — EAGLE-1 recipe also defaults to safetensors + consolidated (train_eagle1.py:207-210). Updated the tutorial with the same clarifying comment.

File: fern/versions/nightly/pages/guides/speculative/eagle.mdx (hunk 3 above)
Comment (HuiyingLi) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:310]: Thank you @khazic , I ran into erro with safetensor+consolidated during training saving ckpt. It failed because the drafter doesn't have the HF state-dict metadata. The sglang exprt code works with draft_model.pt. But during training this path doesn't seem to work. Could you please verify? Thank you so much, I appreciate it!

Comment (HuiyingLi) [reply fern/versions/nightly/pages/guides/speculative/eagle.mdx:310]: @khazic Thank you for checking.  Missed your comments earlier. I will double check on my end.

**Review by HuiyingLi — APPROVED** [review]
Thank you for the amazing tutorial!
